### ISSUE TITLE
How to modify the environment of script compiled in non-standalone mode?

### ISSUE THREAD

**Issue by brupelo** [Nuitka/Nuitka#149]
Hi,

Experiment: Consider this little [hello world example](https://www.dropbox.com/s/0fvnzacb8grvrpr/hello_world.7z?dl=1). If you uncompress and run `x.bat` you'll see how `x.exe` will run ok. That example is using embedded python with `x.py` compiled in non-standalone mode.

`x.py`:

    import os

    os.environ["PATH"] = f"python;{os.environ['path']}"
    os.environ["PYTHONPATH"] = f"python"
    print('-'*80)
    print(os.environ["PATH"])
    print(os.environ["PYTHONPATH"])
    print('-'*80)
    print('hello world')

    import encodings

    print(encodings)

Goal: I'm trying to figure out how to run `x.exe` without relying on any batch file though. As you can see, I've already tried modifying the environment from `x.py` itself but that won't work... so my question would be, is there any way to modify `x.exe` environment without using any additional batch files? It'd be great if there was any, by doing so I'd be able to have a final tidy package with all my external libraries living on their own folders, embedded python living on their own, etc...

If anyone knows a proper way to improve this little experiement please let me know.

Ps. I've googled quite a lot about this stuff but haven't found any satisfactory solution so far.

Thanks in advance.
Bruno.

**Comment by brupelo** [comment]
Tsk! No luck so far... I've tried quite a lot of options to reach my goal... unfortunately none of them worked, the last attempt was tweaking `MainProgram.c`:

    ...
    const char * WinGetEnv(const char * name)
    {
        const DWORD buffSize = 65535;
        static char buffer[buffSize];
        if (GetEnvironmentVariableA(name, buffer, buffSize))
        {
            return buffer;
        }
        else
        {
            return 0;
        }
    }

    #ifdef _NUITKA_WINMAIN_ENTRY_POINT
    int __stdcall WinMain( HINSTANCE hInstance, HINSTANCE hPrevInstance, char* lpCmdLine, int nCmdShow )
    {
    #if defined(__MINGW32__) && !defined(_W64)
        /* MINGW32 */
        int argc = _argc;
        char** argv = _argv;
    #else
        /* MSVC, MINGW64 */
        int argc = __argc;
        char** argv = __argv;
    #endif
    #else
    int main( int argc, char **argv )
    {
        char path[65536];
        char pythonpath[65536];
        sprintf(path, "%s;%s", "python", WinGetEnv("PATH"));
        sprintf(pythonpath, "python;python36.zip");
        SetEnvironmentVariable("PATH", path);
        SetEnvironmentVariable("PYTHONPATH", pythonpath);
        printf("PATH==> %s\n", path);
        printf("PYTHONPATH==> %s\n", pythonpath);
    ...

And then using [/DELAYLOAD](https://msdn.microsoft.com/en-us/library/151kt790.aspx) such as `/DELAYLOAD:python36.dll` in my head that idea would work... but no, by doing that you'll get a linker error as python36.dll is exporting data:

    LINK : fatal error LNK1194: cannot delay-load 'python36.dll' due to import of data symbol '__imp___Py_NoneStruct'; link without /DELAYLOAD:python36.dll

At this point it feels like I've reached a dead-point... I've run out of options... not sure if this problem is solvable... the main idea I've got in mind was I'd avoid the whole overhead of the expensive standalone compilation and I'd end up with a final package containing a little executable, a nice embedded working out of the box embedded python and a clean final layout such as:

![test](https://dl.dropboxusercontent.com/s/jf3jnjc51mro95i/explorer_2018-09-24_17-41-10.png)

but so far the only working solutions I've found working were this:

![ugly1](https://dl.dropboxusercontent.com/s/gag9jpi6g81giii/explorer_2018-09-24_17-43-54.png)

or this:

![ugly2](https://dl.dropboxusercontent.com/s/gbjm64dphd4exoj/explorer_2018-09-24_17-44-33.png)

And I don't like neither of them... tsk! :/

**Comment by kayhayen** [comment]
So are you trying to re-invent minimal standalone Python distributions?

There are a few like those already, MiniConda somes to mind now, but I recall seeing a presentation from people who had one like that for at least Linux alikes. I think CPython might be a relatively bad choice for these things.

As for linking against CPython, I totally dislike what's going on for Windows there. For accelerated and standalone binaries on other platforms, I can specifify "rpath" and make the binary load from binary relative paths any DLL that I want to. On Windows, only binary directory will work in my probing.

Using the ".LIB" of Python is to blame mostly there, because it immediately loads the DLL from well, the configured spots. I don't think the ".bat" can influence that. Maybe manifests embedded in a binary could, but I believe that has limitations too, like absolute paths, and it's cumbersome to learn all that just to find it not working, although I partially tried and failed.

Having our own ".LIB" and loading CPython DLL dynamically would totally solve the issue. A binary once running can change its load paths. It could add what I am thinking of the rpath at run time in C code, rather than through the linker. Not sure what it means for ultimate performance, but it annoys the hell of of me, that for uninstalled Python I need to copy the Python.DLL along the binary to it to work.

**Comment by kayhayen** [comment]
I am not going to make guesses at what you modified and then doesn't work, I would you at code however.

For static compilation, Nuitka seems mostly fit, but there #95 to solve, which actually ought to be very easy, but platform/toolset specific I bet. Static Python linkage would solve the problem, but as you note, CPython doesn't support it on Windows.

One day a performance comparison static/dynamic ought to be made, but my expectation is that going through the trampolines for DLL calls must be slower by a fair margin, first it should be implemented though.

I do not care about cached C files yet, but you know what already.

Maybe one day somebody will script a bootstrap for a static CPython build on Windows, but maybe that won't even be supported, unsure. Would have to check if Anaconda does it. Because I think they recently stopped being dynamic on non-Windows. Maybe eventually they are going to be it, which would make for a recommendation of it.

Yours,
Kay

**Comment by kayhayen** [comment]
My check for static Python on Windows revealed that there are limited projects that are doing it, but fails to include important stuff. So it basically won't exist. I have yet to check what Anaconda on Windows is doing, but my thinking is shared library for Python DLL.

As for loading the DLL delayed, it would be great, but it seems to involve an amount of work that nobody is going to do.

My question to you would be if there is anything going on for this alternative standalone distribution method, and if you still pursue it. I think it seems like a failed attempt?

**Comment by kayhayen** [comment]
Hm, but it you were willing to provide creating the `.bat` (or rather `.cmd` ?) change as PR, that would be good I think. There is another issue where somebody suggested something like this, and if e.g. `main.exe` moved into the folder, and all it did, was calling it from there, the subfolder would be achieved easily, and then it would be tidy, no even ``PATH`` needed, right?

Or do you insist on splitting `python` and other parts really?

I for my part agree with the dislike of single file solution, esp. since data files are always going to be an issue right after that, but I would fancy a tidier distribution any time. The idea of a batch file to launch the exe never occurred the me.

Basically:

```
normal.dist/normal.exe
normal.dist/crapfiles

```
->
```
normal.dist/normal.cmd
normal.dist/normal/normal.exe
normal.dist/normal/crapfiles
```

On non-Windows, the "normal.bin" could just like to relative directory rpath and be good. It wouldn't need a batch file, and it would be consistent.

Taking that on?

**Comment by kayhayen** [comment]
Nah, actually the `.cmd` is only there to trigger the execution of what really remains the moster folder with the data files. Do you want the data files near the launcher? For data files Nuitka currently puts there, that is never the case, like certifi or nose version file, these are all just crap files.

What I was also thinking, why not `.lnk` file pointing to a relative path, if Windows allows for it. And if you want the data directory, well more `.lnk` files I guess.

That feels easy, but might take pywin32 or something heavy in comparison, but maybe only something that ctypes could do too. Then it should be clean in my eyes.

Yours,
Kay

**Comment by kayhayen** [comment]
Used the wrong button apparently, sorry

**Comment by kayhayen** [comment]
Lack of interest, closing until somebody wants to work on it.

**Comment by brupelo** [comment]
I'd forgotten this issue was still opened... certainly it's an interesting one, usually all my deployments look like this nowadays:

![showcase](https://dl.dropboxusercontent.com/s/fwnqw6nw1mw67e3/explorer_2019-07-14_15-36-38.png)

It's not an ideal solution relying on the batch as I'd commented few months ago but at least this way I can separate clearly the minimal exe from the environment which is great for updates. Users will only need to update/download a "small" exe without downloading hundreds of mbs of python packages / deps over and over :/

Anyway, certainly this has always been a really interesting issue as the ideal solution would be not using these batch files at all.

Sad thing is this issue has been open almost 1 year and nobody was interested on it so I guess I was the only one concerned about it afterall hehe ;)

**Comment by MKdays** [comment]
It's not the ultimate solution, but how about this?

Create .bat

```
SET mypath=%~dp0
start %mypath%Package\ExcelTool.exe
```

Convert .bat to exe by bat exe converter (no console mode)
![image](https://user-images.githubusercontent.com/64838226/101911314-88a4c400-3c03-11eb-96ba-41af77a1ca85.png)

It works well and the layout is simple, but I don't think this is a good way.

**Comment by MKdays** [comment]
So I thought about it more deeply.

After making the following modifications in the mainprogram.c, at least .dll files can be placed in subfolders.

```
#if defined(_WIN32)
    // SetDllDirectoryW(getBinaryDirectoryWideChars());
    SetDllDirectory("./Lib/");
```

![image](https://user-images.githubusercontent.com/64838226/101912890-aecb6380-3c05-11eb-9abf-7d3b9a0db19b.png)

But I couldn't find a way to put the .pyd file in the subfolder.

**Comment by brupelo** [comment]
@MKdays Haven't played with Nuitka since opening this issue so I can't tell if the way you're suggesting there would be the recommended way to go nowadays. Now, If I'd seen your proposed solution when opened the issue I'd probably use it myself with slight modifications:

- I'd probably use that bat->exe tool your suggesting just once, I'd reuse the generated exe amongst all my projects (ie: in your particular case I'd rename Run.exe by ExcelTool.exe)
- Inside Package folder the exe would be called the same way amongst projects (ie: run.exe, main.exe or whatever generic name you can think of)

That way you could reuse the bat->exe loader amongst different projects easily without having to spawn the bat->exe conversion each time... Can't think of any major problem your initial solution would give you out of my head tbh other than having the **child->parent process relationsip** created after spawning the loader which obviously is not ideal :/

Anyway, lovely someone is looking at this old one :) , having a final clean final package is definitely a nice thing to have

**Comment by kayhayen** [comment]
Are you people aware of the (albeit experimental) new feature --onefile, also available on Windows? Not sure what was tried to achieve here, but cleaner file system layout for dist folder doesn't matter as much anymore, does it.
